go/vt/vtadmin/cache/cache.go

```diff
@@ -0,0 +1,262 @@
+	// NoExpiration is used to create caches that do not expire items by
+	// default.
+	// See https://pkg.go.dev/github.com/patrickmn/go-cache@v2.1.0+incompatible#pkg-constants.
+	NoExpiration = cache.NoExpiration
+
+	// DefaultBackfillEnqueueWaitTime is the default value used for waiting to
+	// enqueue backfill requests, if a config is passed with a non-positive
+	// BackfillEnqueueWaitTime.
+	DefaultBackfillEnqueueWaitTime = time.Millisecond * 50
+	// DefaultBackfillRequestTTL is the default value used for how stale of
+	// backfill requests to still process, if a config is passed with a
+	// non-positive BackfillRequestTTL.
+	DefaultBackfillRequestTTL = time.Millisecond * 100
+)
+
+// Config is the configuration for a cache.
+type Config struct {
+	// DefaultExpiration is how long to keep Values in the cache by default (the
+	// duration passed to Add takes precedence). Use the sentinel NoExpiration
+	// to make Values never expire by default.
+	DefaultExpiration time.Duration `json:"default_expiration"`
+	// CleanupInterval is how often to remove expired Values from the cache.
+	CleanupInterval time.Duration `json:"cleanup_interval"`
+
+	// BackfillRequestTTL is how long a backfill request is considered valid.
+	// If the backfill goroutine encounters a request older than this, it is
+	// discarded.
+	BackfillRequestTTL time.Duration `json:"backfill_request_ttl"`
+	// BackfillRequestDuplicateInterval is how much time must pass before the
+	// backfill goroutine will re-backfill the same key. It is used to prevent
+	// multiple callers queuing up too many requests for the same key, when one
+	// backfill would satisfy all of them.
+	BackfillRequestDuplicateInterval time.Duration `json:"backfill_request_duplicate_interval"`
+	// BackfillQueueSize is how many outstanding backfill requests to permit.
+	// If the queue is full, calls to EnqueueBackfill will return false and
+	// those requests will be discarded.
+	BackfillQueueSize int `json:"backfill_queue_size"`
+	// BackfillEnqueueWaitTime is how long to wait when attempting to enqueue a
+	// backfill request before giving up.
+	BackfillEnqueueWaitTime time.Duration `json:"backfill_enqueue_wait_time"`
+}
+
+// Cache is a generic cache supporting background fills. To add things to the
+// cache, call Add. To enqueue a background fill, call EnqueueBackfill with a
+// Keyer implementation, which will be passed to the fill func provided to New.
+//
+// For example, to create a schema cache that can backfill full payloads (including
+// size aggregation):
+//
+//		var c *cache.Cache[BackfillSchemaRequest, *vtadminpb.Schema]
+//		c := cache.New(func(ctx context.Context, req BackfillSchemaRequest) (*vtadminpb.Schema, error) {
+//			// Fetch schema based on fields in `req`.
+//			// If err is nil, the backfilled schema will be added to the cache.
+//			return cluster.fetchSchema(ctx, req)
+//		})
+//
+type Cache[Key Keyer, Value any] struct {
+	cache     *cache.Cache
+	fillcache *cache.Cache
```

Comment:
I don't fully understand the behavior of `fillcache`. It seems to be keeping track of the last time we enqueued a fill, so it doesn't happen too often, but if that's the only purpose, it doesn't need to be a cache. It can simply be a `map[Key]time.Time`.

Reply:
(just make sure to `delete` from the map after each read)

Reply:
yep, you have the intended purpose exactly right! i was using the cache because it's threadsafe, if i take your suggestion i'll need to manually manage a mutex for it (which is fine, i think in hindsight i was just being a little lazy and "well, i've already got caches on the brain")

Reply:
> (just make sure to `delete` from the map after each read)

I actually don't think so, but double check my thinking:

Imagine your backfill queue was `[k1, k1, k1]` and your wait interval was 10 minutes.

On the first loop, you fill `k1`, and record `time.Now()`.
On the second loop, you try to fill `k1` again, but it's been less than 10 minutes, so you drop the request.
If you delete the key from `fillcache`, the third loop would then re-fill `k1` even though it hasn't been 10 minutes.

So I think the map has to grow forever. The only reason to delete would be if, when you read, the fill time is outside the wait period, but then you're going to re-fill the cache which will just put the key right back in the map. The only "space" you would save is if your re-fills are failing (i.e. `fillFunc` returns `err != nil`, which _should_ happen (ideally) rarely.

Reply:
Huh, wouldn't that also apply to the cache though?

Reply:
yeah in most practical cases, probably (but we also don't explicitly `delete` from the cache currently, we just expire things)

Reply:
well, okay, _we_ never explicitly delete things, but if you specify a non-zero `CleanupInterval`, then `go-cache` runs a background thread that deletes things that are expired, so infrequently-fetched schemas will get purged over time